js/src/trace.ts

```diff
@@ -0,0 +1,240 @@
+   * Returns an empty array when no experiment is associated with the context.
+   *
+   * First checks the local span cache for recently logged spans, then falls
+   * back to BTQL API if not found in cache.
+   */
+  async getSpans({ spanType }: { spanType?: string[] } = {}): Promise<any[]> {
+    if (!this.experimentId) {
+      return [];
+    }
+
+    const state = _internalGetGlobalState();
+    if (!state) {
+      return [];
+    }
+
+    // Try local cache first
+    const cachedSpans = state.spanCache.getByRootSpanId(this.rootSpanId);
+    if (cachedSpans && cachedSpans.length > 0) {
+      let spans = cachedSpans.filter(
+        (span) => span.span_attributes?.type !== "score",
+      );
+
+      // Apply spanType filter if specified
+      if (spanType && spanType.length > 0) {
+        spans = spans.filter((span) =>
+          spanType.includes(span.span_attributes?.type ?? ""),
+        );
+      }
+
+      return spans.map((span) => ({
+        input: span.input,
+        output: span.output,
+        metadata: span.metadata,
+        span_id: span.span_id,
+        span_parents: span.span_parents,
+        span_attributes: span.span_attributes,
+      }));
+    }
+
+    // Cache miss - fall back to BTQL
+    await this.ensureSpansReady();
+
+    await state.login({});
+
+    const query = `
+      from: experiment('${this.experimentId}')
+      | filter: root_span_id = '${this.rootSpanId}' ${spanType ? `AND span_attributes.type IN ${JSON.stringify(spanType)}` : ""}
+      | select: *
+      | sort: _xact_id asc
+    `;
+
+    for (let attempt = 0; attempt < MAX_FETCH_RETRIES; attempt++) {
+      const response = await state.apiConn().post(
+        "btql",
+        {
+          query,
+          use_columnstore: true,
+          brainstore_realtime: true,
+        },
+        { headers: { "Accept-Encoding": "gzip" } },
+      );
+
+      const payload = await response.json();
+      const rows = payload?.data ?? [];
+      const freshness = payload?.freshness_state;
+      const isFresh =
+        freshness?.last_processed_xact_id != null &&
+        freshness?.last_processed_xact_id ===
+          freshness?.last_considered_xact_id;
+
+      if ((rows.length > 0 && isFresh) || attempt === MAX_FETCH_RETRIES - 1) {
+        return rows
+          .filter((row: any) => row.span_attributes?.type !== "score")
+          .map((row: any) => ({
+            input: row.input,
+            output: row.output,
+            metadata: row.metadata,
+            span_id: row.span_id,
+            span_parents: row.span_parents,
+            span_attributes: row.span_attributes,
+          }));
+      }
+
+      const backoff =
+        INITIAL_RETRY_DELAY_MS * Math.pow(2, Math.min(attempt, 3));
+      await sleep(backoff);
+    }
+
+    return [];
+  }
+
+  /**
+   * Fetch the thread of messages for this trace.
+   *
+   * @experimental This method is experimental and may change in the future.
+   */
+  async getThread() {
```

Comment:
not sure i follow this. i think ideally we should be using the import method in lingua.